---
title: "Figure 1"
author: "Anna Malkowska"
date: "2025-10-07"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(DiffBind)
library(tidyverse)
library(biomaRt)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(ChIPseeker)
#library(clusterProfiler)
options(future.globals.maxSize = 900 * 1024^2)
```

```{r}
#getwd()
#setwd to path where you have bam and peak files and table file
setwd("E:/For paper/DiffBind/new_rerun/new_access")

#read in the metadata
#create metadata based on the table shown in the diffbind vignette
table<- read.csv("epNSCvsqNSC_table.csv")
db.ep <- dba(sampleSheet=table)
```

```{r}
# minOverlap is the proportion of samples in which the peak must be found to be included (if 1 or more confusingly it becomes the number of samples)
# fragmentSize is used as the length of the reads and they will be extended, size from BAM file used if set to 0

# summits - TRUE means that peaksets are unaffected by use of summit calculations
db.ep <- dba.count(db.ep, bSubControl=TRUE, minOverlap=1, fragmentSize = 0, summits = TRUE)

# normalise (default by sequencing depth)
db.ep <- dba.normalize(db.ep)
# model the contrast
db.ep <- dba.contrast(db.ep, reorderMeta=list(Tissue="epNSC"))
#analyse!!
db.ep <- dba.analyze(db.ep)

db.ep.DB <- dba.report(db.ep, bAll = T, th = 1)

#read in the metadata
#create metadata based on the table shown in the diffbind vignette
table<- read.csv("lpNSCvsqNSC_table.csv")
db.lp <- dba(sampleSheet=table)

#plot heatmap 
#plot(db)
#dev.off()

# minOverlap is the proportion of samples in which the peak must be found to be included (if 1 or more confusingly it becomes the number of samples)
# fragmentSize is used as the length of the reads and they will be extended, size from BAM file used if set to 0
# summits - TRUE means that peaksets are unaffected by use of summit calculations
db.lp <- dba.count(db.lp, bSubControl=TRUE, minOverlap=1, fragmentSize = 0, summits = TRUE)

# normalise (default by sequencing depth)
db.lp <- dba.normalize(db.lp)
# model the contrast
db.lp <- dba.contrast(db.lp, reorderMeta=list(Tissue="lpNSC"))
#analyse!!
db.lp <- dba.analyze(db.lp)

db.lp.DB <- dba.report(db.lp, bAll = T, th = 1)
```

Figure 1C,D
```{r}
db.ep.DB.df <- as.data.frame(db.ep.DB)
db.lp.DB.df <- as.data.frame(db.lp.DB)

ep_df <- filter(db.ep.DB.df, FDR < 0.05)
lp_df <- filter(db.lp.DB.df, FDR < 0.05)

ep_df$name <- ifelse(test = ep_df$Fold > 0, yes = "qNSCs", no = "epNSCs")
lp_df$name <- ifelse(test = lp_df$Fold > 0, yes = "qNSCs", no = "lpNSCs")

ggplot(ep_df,
       aes(x=name, fill = name, show.legend = F)) +
       geom_bar(show.legend = F)+
  theme_bw(base_size = 7)+
  theme(axis.title.x = element_blank())+
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, hjust=0.5))+
  ylab("Number of peaks with FDR<0.05")+
  scale_y_continuous(limits = c(0,60000), breaks= scales::pretty_breaks(n=6))+
  scale_fill_manual(values=c("#D9463B", "#5C68AE"))

ggsave(filename = "epNSCvsqNSCFDR_3001.pdf",width = 25, height = 40, unit = "mm", dpi = 300)


ggplot(lp_df,
       aes(x=reorder(name,name,
                     function(x)-length(x)), fill = name)) +
       geom_bar(show.legend = F)+
  theme_bw(base_size = 7)+
  theme(axis.title.x = element_blank())+
  theme(axis.text.x = element_text(angle = 30, vjust = 0.5, hjust=0.5))+
  ylab("Number of peaks with FDR<0.05")+
  xlab("Condition") +
  scale_y_continuous(limits = c(0,60000), breaks= scales::pretty_breaks(n=6))+
  scale_fill_manual(values=c("#D9463B", "#5C68AE"))

ggsave(filename = "lpNSCvsqNSCFDR_new.pdf",width = 25, height =40, unit = "mm", dpi = 300)
```

Figure 1D
Custom Volcano 
```{r}
dba.plotVenn(db.lp, contrast=1, bDB=TRUE,bGain=TRUE, bLoss=TRUE, bAll=F)
db.lp.DB <- dba.report(db.lp, bAll = T, th = 1)

vol.data <- as.data.frame(db.lp.DB)
vol.data <- dplyr::select(vol.data, Fold, FDR)
#vol.data <- filter(vol.data, vol.data$FDR <0.01)

#log the FDR
vol.data$logFDR <- -log10(vol.data$FDR)
vol.data$FC <- -(vol.data$Fold)

#plot with conditional colour&labels
volc_lp <- ggplot(vol.data, aes(x=FC, y=logFDR, colour = 
                                  ifelse(FDR <0.05 & Fold <0,"#DA463B", ifelse(FDR <0.05,"#868D9D", "#5D69B0"))))+
  geom_point(alpha=0.1, size=1)+
  xlab("log2 Fold Change")+
  ylab("-log10(FDR)")+
  geom_vline(xintercept = 1, linetype = "dashed", linewidth=0.5, color = "grey")+
  geom_vline(xintercept = -1, linetype = "dashed", linewidth=0.5, color = "grey")+
  geom_hline(yintercept = 1.29, linetype = "dashed", linewidth=0.5, color = "grey")+
  theme_bw(base_size = 8)+
  ylim(0,60)+
  xlim(-10,10)+
  theme(legend.position = 'none')+
  scale_colour_manual(values=c("#868D9D","#5C68AE","#DA463B" ))


volc_lp

ggsave(filename = "Figure1D.pdf",width = 57, height = 40, unit = "mm")
ggsave(filename = "Figure1D.png",width = 57, height = 40, unit = "mm", dpi = 330)
```

Figure 1C 
Custom Volcano 
```{r}
dba.plotVenn(db.ep, contrast=1, bDB=TRUE,bGain=TRUE, bLoss=TRUE, bAll=F)
db.ep.DB <- dba.report(db.ep, bAll = T, th = 1)

vol.data <- as.data.frame(db.ep.DB)
vol.data <- dplyr::select(vol.data, Fold, FDR)
#vol.data <- filter(vol.data, vol.data$FDR <0.01)

#log the FDR
vol.data$logFDR <- -log10(vol.data$FDR)

#changed the highest datapoint because it was giving me trouble
#vol.data <-arrange(vol.data, desc(logFDR))
#vol.data[1,5] <- 200

#plot with conditional colour&labels
#plot with conditional colour&labels
volc_ep <- ggplot(vol.data, aes(x=Fold, y=logFDR, colour = 
                                  ifelse(FDR <0.05 & Fold <0,"#DA463B", ifelse(FDR <0.05,"#868D9D", "#5D69B0"))))+
  geom_point(alpha=0.1, size=1)+
  xlab("log2 Fold Change")+
  ylab("-log10(FDR)")+
  geom_vline(xintercept = 1, linetype = "dashed", linewidth=0.5, color = "grey")+
  geom_vline(xintercept = -1, linetype = "dashed", linewidth=0.5, color = "grey")+
  geom_hline(yintercept = 1.29, linetype = "dashed", linewidth=0.5, color = "grey")+
  theme_bw(base_size = 8)+
  ylim(0,60)+
  xlim(-10,10)+
  theme(legend.position = 'none')+
  scale_colour_manual(values=c("#868D9D","#5C68AE","#DA463B"))
  
volc_ep

ggsave(filename = "Figure1C.pdf",width = 57, height = 40, unit = "mm")
ggsave(filename = "Figure1C.png",width = 57, height = 40, unit = "mm", dpi = 900)
```

Figure 1E
```{r}
txdb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
qNSCvsepNSC.DB <- dba.report(db.ep, bGain = T)
db.anno.ep <- qNSCvsepNSC.DB[["peaks"]][[2]]
db.anno.ep <- makeGRangesFromDataFrame(db.anno.ep)
seqlevelsStyle(db.anno.ep) <- "UCSC"
db.anno.ep <- annotatePeak(db.anno.ep,tssRegion=c(-300,300), TxDb=txdb) 
plot <- plotAnnoBar(db.anno.ep)
plot
plot_df <- plot[["plot_env"]][["p"]][["data"]]



palette <-c("#D9463B", "#E87C89", "#F4991C", "#F4C412", "#F9E498", "#80518E", "#A174B3", "#5C68AE","#289A8D")
palette <- rev(as.vector(palette))

plot_df$Feature <- factor(plot_df$Feature, ordered = TRUE, levels = plot_df$Feature)
plot_df$Feature <- forcats::fct_rev(plot_df$Feature)
# Stacked + percent

ggplot(plot_df, aes(x='', fill=Feature, y=Frequency)) + 
      geom_bar(stat="identity") + coord_flip() + ggtitle("Feature Distribution")+
      ylab("Percentage")+
      theme_bw() + theme(axis.title.y=element_blank(),
                 axis.text.y=element_blank(),
                  axis.ticks.y=element_blank(), 
                  legend.key.size = unit(2, 'mm'), 
                  axis.text = element_text(size=8),
                   axis.title = element_text(size=8),
                 legend.text=element_text(size=7),
                  legend.title = element_text(size=7),plot.title=element_text(size=8), 
                 plot.margin = margin(t=0,r=0,b=0,l=0), 
                  legend.margin = margin(t=0,r=0,b=0,l=0))+
          scale_fill_manual(values=palette)+ 
        guides(fill = guide_legend(reverse=TRUE)) 
ggsave(filename = "Feat_Distr.pdf",width = 83, height = 36, unit = "mm")        
```

Figure 1F
```{r}
db.anno.ep <- as.data.frame(db.anno.ep)
db.anno.ep <- dplyr::select(db.anno.ep, start, geneId)

db.fc.ep <- qNSCvsepNSC.DB[["peaks"]][[2]]
db.fc.ep <- dplyr::select(db.fc.ep, Start, Fold)
db.fc.ep <- rename(db.fc.ep, "start" = "Start")

db.go.ep <- inner_join(db.anno.ep, db.fc.ep, by="start")

db.go.ep<- db.go.ep %>%
  arrange(desc(geneId), desc(Fold))

db.go.ep <- db.go.ep %>% distinct(geneId, .keep_all = TRUE)
db.go.ep <- dplyr::select(db.go.ep, -start)

db.go.ep<- db.go.ep %>%
  arrange(desc(Fold))

db.go.ep_filt <- filter(db.go.ep, Fold > 2)

###
#BiocManager::install("PANTHER.db")
#install.packages("rbioapi")
library(rbioapi)
org <- rba_panther_info(what = "organisms")
annots <- rba_panther_info(what = "datasets")

#db.go.ep_filt$geneId

enriched <- rba_panther_enrich(genes = db.go.ep_filt$geneId,
                               organism = 7227,
                               annot_dataset = "ANNOT_TYPE_ID_PANTHER_GO_SLIM_BP",
                               cutoff = 0.01)

result_qNSC <- enriched[["result"]]
#write.csv(result_qNSC, "ExtData1_Fig1F.csv")

result_qNSC <- dplyr::filter(result_qNSC, number_in_reference < 500, number_in_list >10)
result_qNSC <- dplyr::filter(result_qNSC, plus_minus == "+")
result_qNSC$term.label <- factor(result_qNSC$term.label, ordered=TRUE, levels = result_qNSC$term.label[order(result_qNSC$fold_enrichment, decreasing=TRUE)])


result_qNSC <- result_qNSC[c(1,6,7,11,14,16),]

result_qNSC$term.label <- sub("(?<=^.{40}).*","...", result_qNSC$term.label, perl = T)

library(MetBrewer)
palette <- met.brewer(name="Signac", n=10)[1:3]
palette <- c("#FDF7E1", palette)

library(ggplot2)
ggplot2::ggplot(result_qNSC) +
  geom_col(aes(fill=fdr, y = term.label,
               x=fold_enrichment))+
  xlab("Fold Enrichment")+
  ylab("Gene Ontology entry")+
  labs(fill='FDR')+
  theme_bw()+
    ggtitle("Most accessible peaks in qNSCs")+
  theme(legend.title = element_text(size=6),plot.title=element_text(size=8, hjust = -0.2),
                    legend.key.size = unit(2, 'mm'),
                  axis.text = element_text(size=7),
                   axis.title = element_text(size=7),
                 legend.text=element_text(size=6),
                 plot.margin = margin(t=0,r=0,b=0,l=0), 
                  legend.margin = margin(t=0,r=0,b=0,l=0))+
  scale_fill_gradientn(colours=palette)

ggsave("Figure1F.pdf",width = 83, height = 36, unit = "mm")
```

